gateway-service/src/main/java/org/zowe/apiml/gateway/security/config/SecurityConfiguration.java

```diff
@@ -174,10 +173,16 @@ protected void configure(HttpSecurity http) throws Exception {
 
             // add filters - login, query, ticket
             .and()
-            .addFilterBefore(loginFilter(authConfigurationProperties.getGatewayLoginEndpoint()), UsernamePasswordAuthenticationFilter.class)
-            .addFilterBefore(x509Filter(authConfigurationProperties.getGatewayLoginEndpoint()), LoginFilter.class)
-            .addFilterBefore(loginFilter(authConfigurationProperties.getGatewayLoginEndpointOldFormat()), UsernamePasswordAuthenticationFilter.class)
-            .addFilterBefore(x509Filter(authConfigurationProperties.getGatewayLoginEndpointOldFormat()), LoginFilter.class)
+
+            .addFilterBefore(new ShouldBeAlreadyAuthenticatedFilter(authConfigurationProperties.getGatewayLoginEndpoint(), handlerInitializer.getAuthenticationFailureHandler()), UsernamePasswordAuthenticationFilter.class)
```

Comment:
Do addFilterBefore shifts the filters?

Reply:
They are added "before" the specified filter, which means that they are processed earlier. The filters are stored in list. In this case, the order of filter processing is loginFilter -> x509AuthenticationFilter -> ShouldBeAlreadyAuthenticatedFilter -> UsernamePasswordAuthenticationFilter.